### ISSUE TITLE
TSX data-* props issue with Stricter Assignability Checks to Unions with Index Signatures

### ISSUE THREAD

**Issue by jraoult** [microsoft/TypeScript#36935]
Seems like the new _Stricter Assignability Checks to Unions with Index Signatures_ in 3.8 makes it impossible to have `data-*` props passed to React components when their props type is based on a union.

Pardon the contrived example below (I tried to reduce as much as possible for clarity sake). For a real case most material UI components props types are using unions and therefor fail the same test.

**TypeScript Version:**  3.8-Beta

**Search Terms:**
data
Stricter Assignability Checks to Unions with Index Signatures

**Expected behavior:**
The `data-` props should be accepted all the time (3.7 behaviour)

**Actual behavior:**
With 3.8 any extra `data-*` will be rejected if the props type is based on a union.

**Related Issues:**
In another repo, but may be this https://github.com/mui-org/material-ui/issues/19536 ?

**Code**
```ts
import React, { ReactElement} from "react";

declare function NotHappy(props: ({ fixed?: boolean } | { value?: number })): ReactElement;
declare function Happy(props: { fixed?: boolean, value?: number }): ReactElement;

const RootNotHappy = () => (<NotHappy data-testid="my-test-id" />)
const RootHappy = () => (<Happy data-testid="my-test-id" />)
```

<details><summary><b>Compiler Options</b></summary>

```json
{
  "compilerOptions": {
    "noImplicitAny": true,
    "strictNullChecks": true,
    "strictFunctionTypes": true,
    "strictPropertyInitialization": true,
    "strictBindCallApply": true,
    "noImplicitThis": true,
    "noImplicitReturns": true,
    "useDefineForClassFields": false,
    "alwaysStrict": true,
    "allowUnreachableCode": false,
    "allowUnusedLabels": false,
    "downlevelIteration": false,
    "noEmitHelpers": false,
    "noLib": false,
    "noStrictGenericChecks": false,
    "noUnusedLocals": false,
    "noUnusedParameters": false,
    "esModuleInterop": true,
    "preserveConstEnums": false,
    "removeComments": false,
    "skipLibCheck": false,
    "checkJs": false,
    "allowJs": false,
    "declaration": true,
    "experimentalDecorators": false,
    "emitDecoratorMetadata": false,
    "target": "Latest",
    "jsx": "Preserve",
    "module": "ESNext"
  }
}
```

</details>

**Playground Link:** [Provided](https://www.typescriptlang.org/play/?target=99&jsx=1&ts=3.8-Beta#code/JYWwDg9gTgLgBAJQKYEMDGMA0cDejUYCiANkiEgHYwC+cAZlBCHAERQEwsDcAsAFD8AJkjTEU7egFcKGYBApwAchBgAJFGDABPABRhGYAM4AuODrx1gADySCA-KYBGECKRQLaAH1xwAbimJJJAc4CkkQRyQoOGoASljTZHQYEjJKGF4+YVFxJCkZGDkFdU1dfQgjUwtrWxDnV1QKbH9A4NMwiKiYhPxk1PIqTP40eUN4BBcYZTUNbTgAXjNYhYA+MwAeaZK5wRQYFABaGCQx4EF5lhAtI5OYA7OWOAB6Fdjh0fHJ7a0FpdWN75wXb7G6nc6Xa7HMb3QSPF5vPhAA)

**Comment by Retsam** [comment]
As a short-term workaround:

```ts
<NotHappy data-testid={"my-test-id" as never} />
```

But I agree, this looks like it should work to me.

**Comment by orta** [comment]
I've heard two reports of this so far, might be worth including in a 3.8 patch also

**Comment by jraoult** [comment]
@orta it would be great. If the fix it is not making it in a patch release for the 3.8 we will have to skip the whole minor. I can imagine that we are not the only ones in this case. This is not uncommon for people to use `data-*` in React, especially peeps using Cypress where it is even [considered a best practice](https://docs.cypress.io/guides/references/best-practices.html#Selecting-Elements).

**Comment by chandlerprall** [comment]
> I've heard two reports of this so far, might be worth including in a 3.8 patch also

This affects Elastic UI as well; similar to Material UI we have a number of components which are unions. If this break is not fixed in 3.8 it not only affects our Elastic&Material UI projects directly but also blocks any downstream consumers from using TypeScript 3.8 + these component libraries.
